File: MonoDll/Interops/Viewport/Renderer.cpp
#include "StdAfx.h"
#include "Renderer.h"

RendererInterop::RendererInterop()
{
	REGISTER_METHOD(ScreenToWorld);
	REGISTER_METHOD(UnProjectFromScreen);

	REGISTER_METHOD(CreateRenderTarget);
	REGISTER_METHOD(DestroyRenderTarget);
	REGISTER_METHOD(SetRenderTarget);

	REGISTER_METHOD(GetViewCamera);

	REGISTER_METHOD(SetCameraMatrix);
	REGISTER_METHOD(GetCameraMatrix);

	REGISTER_METHOD(SetCameraPosition);
	REGISTER_METHOD(GetCameraPosition);

	REGISTER_METHOD(GetCameraFieldOfView);
}

int RendererInterop::UnProjectFromScreen(float sx, float sy, float sz, float &px, float &py, float &pz)
{
	return gEnv->pRenderer->UnProjectFromScreen(sx, sy, sz, &px, &py, &pz);
}

Vec3 RendererInterop::ScreenToWorld(int x, int y)
{
	if (gEnv->pPhysicalWorld)
	{
		float mouseX, mouseY, mouseZ;
		Vec3  camPos = gEnv->pSystem->GetViewCamera().GetPosition();

		gEnv->pRenderer->UnProjectFromScreen((float)x, GetHeight() - (float)y, 0.0f, &mouseX, &mouseY, &mouseZ);
		Vec3 dir = (Vec3(mouseX, mouseY, mouseZ) - camPos).GetNormalizedSafe();

		static ray_hit hit;
		IPhysicalEntity *pPhysEnt = nullptr;

		if (gEnv->pPhysicalWorld->RayWorldIntersection(camPos, dir * gEnv->p3DEngine->GetMaxViewDistance(), ent_all, rwi_stop_at_pierceable | rwi_colltype_any, &hit, 1, pPhysEnt))
			return hit.pt;
	}

	return Vec3(ZERO);
}

int RendererInterop::CreateRenderTarget(int width, int height, ETEX_Format texFormat)
{
	return gEnv->pRenderer->CreateRenderTarget(width, height, texFormat);
}

void RendererInterop::DestroyRenderTarget(int textureId)
{
	gEnv->pRenderer->DestroyRenderTarget(textureId);
}

void RendererInterop::SetRenderTarget(int textureId)
{
	gEnv->pRenderer->SetRenderTarget(textureId);
}

const CCamera *RendererInterop::GetViewCamera()
{
	return &gEnv->pRenderer->GetCamera();
}

void RendererInterop::SetCameraMatrix(CCamera *pCamera, Matrix34 matrix)
{
	pCamera->SetMatrix(matrix);
}

Matrix34 RendererInterop::GetCameraMatrix(CCamera *pCamera)
{
	return pCamera->GetMatrix();
}

void RendererInterop::SetCameraPosition(CCamera *pCamera, Vec3 pos)
{
	pCamera->SetPosition(pos);
}

Vec3 RendererInterop::GetCameraPosition(CCamera *pCamera)
{
	return pCamera->GetPosition();
}

float RendererInterop::GetCameraFieldOfView(CCamera *pCamera)
{
	return pCamera->GetFov();
}

File: CryBrary/Engine/Renderer/Renderer.cs
using System.Runtime.CompilerServices;
using CryEngine.Mathematics;
using CryEngine.Mathematics.Graphics;
using CryEngine.Native;

namespace CryEngine
{
	public static class Renderer
	{

		public static Vector3 ScreenToWorld(int x, int y)
		{
			return RendererInterop.ScreenToWorld(x, y);
		}

		public static int UnProjectFromScreen(float sx, float sy, float sz, out float px, out float py, out float pz)
		{
			return RendererInterop.UnProjectFromScreen(sx, sy, sz, out px, out py, out pz);
		}

		public static int CreateRenderTarget(int width, int height, int flags = 0)
		{
			return RendererInterop.CreateRenderTarget(width, height, flags);
		}

		public static void SetRenderTarget(int id)
		{
			RendererInterop.SetRenderTarget(id);
		}

		public static void DestroyRenderTarget(int id)
		{
			RendererInterop.DestroyRenderTarget(id);
		}
	}
}